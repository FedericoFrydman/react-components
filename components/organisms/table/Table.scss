@import '../../assets/styles/colors.scss';

.lsux-table {
  display: table;
  width: 100%;

  &--border {
    border-bottom: 1px solid $N300;
    border-right: 1px solid $N300;

    & .lsux-table__row__cell {
      border-left: 1px solid $N300;
      border-top: 1px solid $N300;
    }
  }

  &__row {
    display: table-row;
    position: relative;
    width: 100%;

    &--header {
      background-color: $N300;
      font-weight: bold;
    }

    &--odd {
      background-color: $N100;
    }

    &--disabled {
      cursor: not-allowed;
    }

    &--large {
      height: 6rem;
    }

    &--medium {
      height: 4rem;
    }

    &--small {
      height: 3rem;
    }

    &--auto {
      height: 3rem;
    }

    &__cell {
      display: table-cell;
      height: auto;
      padding: 8px;
      text-overflow: ellipsis;

      &--text--left {
        text-align: left;
      }

      &--text--right {
        text-align: right;
      }

      &--text--center {
        text-align: center;
      }

      &--nowrap {
        white-space: nowrap;
      }

      &--content--top {
        vertical-align: top;
      }

      &--content--center {
        vertical-align: middle;
      }

      &--content--bottom {
        vertical-align: bottom;
      }
    }
  }
}
