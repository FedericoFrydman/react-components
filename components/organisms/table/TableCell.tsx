import { forwardRef, HTMLProps, ReactElement, Ref } from 'react';
import cn from 'classnames';

import './Table.scss';

export type TableCellAlign = 'left' | 'right' | 'center';

export interface TableCellProps extends Omit<HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * Content of the cell
   */
  children: ReactElement[] | ReactElement | string | number;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Define the flex-direction
   */
  alignContent?: 'top' | 'center' | 'bottom';
  /**
   * Turns off text wrapping
   */
  noTextWrap?: boolean;
  /**
   * Sets the text align attribute to left/right/center
   */
  textAlign?: TableCellAlign;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const TableCellComponent = (
  { children, classNames = [], noTextWrap = false, textAlign = 'left', alignContent = 'top' }: TableCellProps,
  ref: TableCellProps['ref'],
) => {
  return (
    <div
      className={cn(
        {
          ['lsux-table__row__cell']: true,
          ['lsux-table__row__cell--nowrap']: noTextWrap,
          [`lsux-table__row__cell--text--${textAlign}`]: textAlign,
          [`lsux-table__row__cell--content--${alignContent}`]: alignContent,
        },
        classNames,
      )}
      ref={ref}
    >
      {children}
    </div>
  );
};

export const TableCell = forwardRef<HTMLDivElement, TableCellProps>(TableCellComponent);
