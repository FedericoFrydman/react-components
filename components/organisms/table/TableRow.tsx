import { forwardRef, HTMLProps, ReactElement, Ref } from 'react';
import cn from 'classnames';

import './Table.scss';
import { TableCellProps } from './TableCell';

export type TableRowSize = 'small' | 'medium' | 'large' | 'auto';

export interface TableRowProps extends Omit<HTMLProps<HTMLDivElement>, 'css' | 'size'> {
  /**
   * Column description; if left out the number of children is used
   */
  columns?: string;
  /**
   * Children; typically the cells
   */
  children: ReactElement<TableCellProps>[] | ReactElement<TableCellProps>;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Indicates the row is intended as a header row
   */
  isHeader?: boolean;
  /**
   * Indicates an odd row for alternative background shading
   */
  isOdd?: boolean;
  /**
   * Size of the row
   */
  size?: TableRowSize;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}
const TableRowComponent = (
  {
    isHeader = false,
    size = 'auto',
    classNames = [],
    children,
    isOdd = false,
    columns,
    disabled,
    ...props
  }: TableRowProps,
  ref: TableRowProps['ref'],
) => {
  const childrenLength = Array.isArray(children) ? children.length : 1;

  const classTableRow = 'lsux-table__row';

  return (
    <div
      {...props}
      className={cn(
        {
          [`${classTableRow}`]: true,
          [`${classTableRow}--${size}`]: size,
          [`${classTableRow}--header`]: isHeader,
          [`${classTableRow}--odd`]: isOdd,
          [`${classTableRow}--disabled`]: disabled,
        },
        classNames,
      )}
      style={{ gridTemplateColumns: columns ? columns : '1fr '.repeat(childrenLength), ...props.style }}
      ref={ref}
    >
      {children}
    </div>
  );
};

export const TableRow = forwardRef<HTMLDivElement, TableRowProps>(TableRowComponent);
