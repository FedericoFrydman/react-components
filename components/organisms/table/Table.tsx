import { forwardRef, HTMLProps, ReactElement, Ref } from 'react';
import cn from 'classnames';

import './Table.scss';
import { TableRowProps } from './TableRow';

type RowChildren = [ReactElement<TableRowProps>, ReactElement<TableRowProps>[] | ReactElement<TableRowProps>];

export interface TableProps extends Omit<HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * Indicates that the table should have no border attribute
   */
  noBorder?: boolean;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Children (the TableRows)
   */
  children: ReactElement<TableRowProps>[] | ReactElement<TableRowProps> | RowChildren;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const TableComponent = (
  { noBorder = false, classNames = [], children, ...props }: TableProps,
  ref: TableProps['ref'],
) => {
  const classTable = 'lsux-table';
  const cnTable = cn(classTable, { [`${classTable}--border`]: noBorder }, classNames);

  return (
    <div {...props} className={cnTable} ref={ref}>
      {children}
    </div>
  );
};

/**
 * Tables are used to organize and display structure content or data information in a grid-like format. Tables helps users compare and analyze among a dataset.
 */

export const Table = forwardRef<HTMLDivElement, TableProps>(TableComponent);
