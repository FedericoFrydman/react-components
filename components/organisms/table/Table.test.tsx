import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Table, TableProps } from './Table';
import { TableCell } from './TableCell';
import { TableRow } from './TableRow';

describe('Table Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test-table';
    wrapper = shallow(
      <Table>
        <TableRow key="1" isHeader={true}>
          <TableCell>Column 1</TableCell>
          <TableCell>Column 2</TableCell>
        </TableRow>
        <TableRow key="2">
          <TableCell>Cell 1, 1</TableCell>
          <TableCell>Cell 1, 2</TableCell>
        </TableRow>
        <TableRow key="3" isOdd={true}>
          <TableCell>Cell 2, 1</TableCell>
          <TableCell>Cell 2, 2</TableCell>
        </TableRow>
        <TableRow key="4">
          <TableCell>Cell 3, 1</TableCell>
          <TableCell>Cell 3, 2</TableCell>
        </TableRow>
        <TableRow key="5" disabled isOdd={true}>
          <TableCell>Cell 4, 1 (disabled)</TableCell>
          <TableCell>Cell 4, 2 (disabled)</TableCell>
        </TableRow>
      </Table>,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the Table the provided classNames', () => {
          const table = renderTable({ classNames });

          expect(table).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the snackbar', () => {
      expect(renderTable({ ref })).toEqual(ref.current);
    });
  });

  const renderTable = (props: Omit<TableProps, 'children'> = {}): HTMLElement => {
    render(
      <Table {...props} data-testid={testId}>
        <TableRow key="1" isHeader={true}>
          <TableCell>Column 1</TableCell>
          <TableCell>Column 2</TableCell>
        </TableRow>
        <TableRow key="2">
          <TableCell>Cell 1, 1</TableCell>
          <TableCell>Cell 1, 2</TableCell>
        </TableRow>
        <TableRow key="3" isOdd={true}>
          <TableCell>Cell 2, 1</TableCell>
          <TableCell>Cell 2, 2</TableCell>
        </TableRow>
        <TableRow key="4">
          <TableCell>Cell 3, 1</TableCell>
          <TableCell>Cell 3, 2</TableCell>
        </TableRow>
        <TableRow key="5" disabled isOdd={true}>
          <TableCell>Cell 4, 1 (disabled)</TableCell>
          <TableCell>Cell 4, 2 (disabled)</TableCell>
        </TableRow>
      </Table>,
    );

    return screen.getByTestId(testId);
  };
});
