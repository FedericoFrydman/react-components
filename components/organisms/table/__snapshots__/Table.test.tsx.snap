// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should render the component 1`] = `
<div
  className="lsux-table"
>
  <ForwardRef(TableRowComponent)
    isHeader={true}
    key="1"
  >
    <ForwardRef(TableCellComponent)>
      Column 1
    </ForwardRef(TableCellComponent)>
    <ForwardRef(TableCellComponent)>
      Column 2
    </ForwardRef(TableCellComponent)>
  </ForwardRef(TableRowComponent)>
  <ForwardRef(TableRowComponent)
    key="2"
  >
    <ForwardRef(TableCellComponent)>
      Cell 1, 1
    </ForwardRef(TableCellComponent)>
    <ForwardRef(TableCellComponent)>
      Cell 1, 2
    </ForwardRef(TableCellComponent)>
  </ForwardRef(TableRowComponent)>
  <ForwardRef(TableRowComponent)
    isOdd={true}
    key="3"
  >
    <ForwardRef(TableCellComponent)>
      Cell 2, 1
    </ForwardRef(TableCellComponent)>
    <ForwardRef(TableCellComponent)>
      Cell 2, 2
    </ForwardRef(TableCellComponent)>
  </ForwardRef(TableRowComponent)>
  <ForwardRef(TableRowComponent)
    key="4"
  >
    <ForwardRef(TableCellComponent)>
      Cell 3, 1
    </ForwardRef(TableCellComponent)>
    <ForwardRef(TableCellComponent)>
      Cell 3, 2
    </ForwardRef(TableCellComponent)>
  </ForwardRef(TableRowComponent)>
  <ForwardRef(TableRowComponent)
    disabled={true}
    isOdd={true}
    key="5"
  >
    <ForwardRef(TableCellComponent)>
      Cell 4, 1 (disabled)
    </ForwardRef(TableCellComponent)>
    <ForwardRef(TableCellComponent)>
      Cell 4, 2 (disabled)
    </ForwardRef(TableCellComponent)>
  </ForwardRef(TableRowComponent)>
</div>
`;
